Login form: Add support for second-factor (form-based)

matthiaso opened the pull request:
354738

matthiaso committed:
Login form: Add support for second-factor (form-based)

354738

stephan-merkli commented:
Change commit message to format [affected part]: [what was changed]
Login form: add support for second factor
or
Login form: add support for two-factor authentication

stephan-merkli commented on eclipse-scout-core/src/login/LoginBox.ts:
```diff
@@ -188,15 +223,18 @@ export class LoginBox extends Box {
+        .one('input.resetLoginError', this._resetButtonText.bind(this));
+    }
+
+    (this.$password ? this.$password : this.$token)
```
Use local variable

matthiaso replied:
Code does not exist anymore (in favor of always resetting login completely).

stephan-merkli commented on eclipse-scout-core/src/login/LoginBox.ts:
```diff
@@ -188,15 +223,18 @@ export class LoginBox extends Box {
+      this.$user
+        .val('')
+        .focus();
+      this.$user
```
user should always be reset to. If second factor is invalid, login process should start at the begining again, not only with 2. factor (?)

paolobazzi replied:
Agree

matthiaso replied:
Also reset backend session

matthiaso replied:
Fixed in f580ed9a78b2af88bdf738a08859d722b6adfd5a

stephan-merkli commented on org.eclipse.scout.rt.server.commons/src/main/java/org/eclipse/scout/rt/server/commons/authentication/FormBasedAccessController.java:
```diff
@@ -117,6 +124,69 @@ protected boolean handleAuthRequest(HttpServletRequest request, HttpServletRespo
+    }
+
+    // OWASP: force a new HTTP session to be created.
+    ServletFilterHelper helper = BEANS.get(ServletFilterHelper.class);
```
Create own method, similar code as in 117-123, add principal as parameter

matthiaso replied:
See f580ed9a78b2af88bdf738a08859d722b6adfd5a, code is not so similar anymore as for two factor authentication session is now always reset. -> No own method created (yet).

stephan-merkli commented on org.eclipse.scout.rt.ui.html/src/main/resources/org/eclipse/scout/rt/ui/html/texts/Texts.properties:
```diff
@@ -161,6 +161,7 @@ ui.TableRowCount0=No row
 ui.TableRowCount1=One row
 ui.TableRowCount={0} rows
 ui.TileView=Tile view
+ui.Token=Token
```
Required in UiTextContributor? Not sure, login box behaves differently.

paolobazzi replied:
All text of scout.rt.ui.html are provided automatically to UI by org.eclipse.scout.rt.ui.html.UiTextContributor

stephan-merkli commented on org.eclipse.scout.rt.server.commons/pom.xml:
```diff
@@ -28,6 +28,10 @@
       <groupId>org.eclipse.scout.rt</groupId>
       <artifactId>org.eclipse.scout.rt.shared</artifactId>
     </dependency>
+    <dependency>
```
Would require Jackson dependency in .app too. Omit and create JSON manually for this simple case?

paolobazzi replied:
Jackson dependency should be available on all .app modules.

rt.shared has already a (transitive) dependency to rt.dataobject through rt.security, therefore I suggest to remove this dependency here.

matthiaso replied:
Removed dependency

matthiaso commented:
> Change commit message to format [affected part]: [what was changed] Login form: add support for second factor or Login form: add support for two-factor authentication

- [x] Changed?

paolobazzi commented on org.eclipse.scout.rt.server.commons/src/main/java/org/eclipse/scout/rt/server/commons/authentication/FormBasedAccessController.java:
```diff
@@ -87,13 +91,11 @@ public void destroy() {
    */
   @SuppressWarnings({"squid:RedundantThrowsDeclarationCheck", "DuplicatedCode"}) // required so that overriding subclasses can throw ServletExceptions
   protected boolean handleAuthRequest(HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException {
```
A unit test covering several cases for handleAuthRequest() would be nice ;-)
E.g. using mocked request and expectation on response?

matthiaso replied:
Fixed in c23066ab4cc592f8bef8ff6d062c5ab7fd48d8a0

paolobazzi reviewed (approved):
Looks good, see inputs for minor improvements.

matthiaso commented:
Changes since last review: https://github.com/eclipse-scout/scout.rt/compare/044fa9a77240c3c6c00d2660766b8d600ea3bb5d..f580ed9a78b2af88bdf738a08859d722b6adfd5a

matthiaso commented:
Changes since last review: https://github.com/eclipse-scout/scout.rt/compare/044fa9a77240c3c6c00d2660766b8d600ea3bb5d..c2fe96c0ff036a624eebaaee2060ce1777e1e8e1

matthiaso commented:
Changes since last review https://github.com/eclipse-scout/scout.rt/compare/044fa9a77240c3c6c00d2660766b8d600ea3bb5d..c23066ab4cc592f8bef8ff6d062c5ab7fd48d8a0

matthiaso commented:
Still use https://github.com/eclipse-scout/scout.rt/compare/044fa9a77240c3c6c00d2660766b8d600ea3bb5d..c23066ab4cc592f8bef8ff6d062c5ab7fd48d8a0 for comparision, last change only rebase

matthiaso commented:
Still use https://github.com/eclipse-scout/scout.rt/compare/044fa9a77240c3c6c00d2660766b8d600ea3bb5d..c23066ab4cc592f8bef8ff6d062c5ab7fd48d8a0 for comparision, last change only rebase
